### DISCUSSION TITLE
slower transfers after network switch

### DISCUSSION THREAD

**Discussion by maoxing** [curl/curl#13120] [Q&A] [answered]
lib version: 7.52.1

i did this:

multi handle:

    /* init curl content */
    CURLM* curlmHandle = curl_multi_init();

    /* never do more than 15 connections */
    curl_multi_setopt(curlmHandle, CURLMOPT_MAX_TOTAL_CONNECTIONS, 15L); 

per easy handle:

    /* enable TCP keep-alive for this transfer */
    curl_easy_setopt(handle, CURLOPT_TCP_KEEPALIVE, 1L);

    /* keep-alive idle time to 120 seconds */
    curl_easy_setopt(handle, CURLOPT_TCP_KEEPIDLE, 120L);

    /* interval time between keep-alive probes: 60 seconds */
    curl_easy_setopt(handle, CURLOPT_TCP_KEEPINTVL, 60L);

In my network environment, the normal download speed is about 1Mb/s-3Mb/s. When I switch networks during the download process, the download speed may drop for a period of time. It is only about kb level speed, and the download will not time out. It will always download at a low speed, I did not set the option CURLOPT_MAX_RECV_SPEED_LARGE

some verbose:

< HTTP/1.1 206 Partial Content
< Server: Tengine
< Content-Type: application/x-xar
< Content-Length: 52325
< Connection: keep-alive
< Date: Wed, 13 Mar 2024 23:17:59 GMT
< x-oss-request-id: 65F23427B899983930A71822
< x-oss-cdn-auth: success
< Accept-Ranges: bytes
< ETag: "C2F281236EBF6704F7F92D4891B7D44E-400"
< Last-Modified: Tue, 12 Mar 2024 02:14:13 GMT
< x-oss-object-type: Multipart
< x-oss-hash-crc64ecma: 14464167936175625912
< x-oss-storage-class: Standard
< x-oss-server-side-encryption: AES256
< x-oss-server-time: 231
< Ali-Swift-Global-Savetime: 1710371879
< Via: cache23.l2cn3158[0,0,206-0,H], cache48.l2cn3158[1,0], vcache3.cn6344[0,0,206-0,H], vcache5.cn6344[1,0]
< Age: 15573
< X-Cache: HIT TCP_MEM_HIT dirn:10:83212175 mlen:0
< Content-Range: bytes 1144464798-1144517122/9004989359
< X-Swift-SaveTime: Thu, 14 Mar 2024 02:11:37 GMT
< X-Swift-CacheTime: 2581582
< Timing-Allow-Origin: *
< EagleId: ab0f249917103874520326247e
<
* No connections available in cache
* No connections available.
* No connections available in cache
* No connections available.
* No connections available in cache
* No connections available.
* No connections available in cache
* No connections available.
* No connections available in cache
* No connections available.
* No connections available in cache
* No connections available.
* No connections available in cache
* No connections available.
* No connections available in cache
* No connections available.
* No connections available in cache

* No connections available.
* Found bundle for host xxxxxxxx: 0x1aae1732280 [serially]
* No connections available in cache
* No connections available.
* Found bundle for host xxxxxxxxx: 0x1aae1732280 [serially]
* No connections available in cache
* No connections available.

I don't know why? Who can give me some advice?

**Comment by bagder** [comment, marked as answer]
> When I switch networks during the download process, the download speed may drop for a period of time

What does this mean? Switching networks could imply that you get a new IP address, and that would then (probably) totally ruin your ability to continue a transfer. But apparently your transfers can continue, so it is hard to understand what this network switch is.

But: curl uses the TCP/IP stack with the normal API calls offered by your system. It does not really know nor care about network specifics and as long as the functions keep returning data correctly, it will continue the transfer. If the data is delivered slower to curl, then it will report a slower transfer speed. I don't think curl can do much to magically make things go faster. It is the network stack in your kernel that does most of that.
